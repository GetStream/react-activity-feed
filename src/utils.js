// @flow
import * as React from 'react';
import URL from 'url-parse';
import anchorme from 'anchorme';
import _truncate from 'lodash/truncate';
import twitter from 'twitter-text';
import type { UserResponse } from './types';
import Dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import minMax from 'dayjs/plugin/minMax';
import type { Renderable, RenderableButNotElement, FileLike } from './types';
Dayjs.extend(utc);
Dayjs.extend(minMax);

// import type { UserResponse } from 'getstream';

export function humanizeTimestamp(
  timestamp: string | number,
  tDateTimeParser: (input?: string | number) => Function,
): string {
  let time;
  // Following calculation is based on assumption that tDateTimeParser()
  // either returns momentjs or dayjs object.

  // When timestamp doesn't have z at the end, we are supposed to take it as UTC time.
  // Ideally we need to adhere to RFC3339. Unfortunately this needs to be fixed on backend.
  if (
    typeof timestamp === 'string' &&
    timestamp[timestamp.length - 1].toLowerCase() === 'z'
  ) {
    time = tDateTimeParser(timestamp);
  } else {
    time = tDateTimeParser(timestamp).add(
      Dayjs(timestamp).utcOffset(),
      'minute',
    ); // parse time as UTC
  }

  const now = tDateTimeParser();
  return time.from(now);
}

export const smartRender = (
  ElementOrComponentOrLiteral: Renderable,
  props?: {},
  fallback?: Renderable,
) => {
  if (ElementOrComponentOrLiteral === undefined) {
    ElementOrComponentOrLiteral = fallback;
  }
  if (React.isValidElement(ElementOrComponentOrLiteral)) {
    // Flow cast through any, to make flow believe it's a React.Element
    return ((ElementOrComponentOrLiteral: any): React.Element<any>);
  }

  // Flow cast through any to remove React.Element after previous check
  const ComponentOrLiteral = ((ElementOrComponentOrLiteral: any): RenderableButNotElement);
  if (
    typeof ComponentOrLiteral === 'string' ||
    typeof ComponentOrLiteral === 'number' ||
    typeof ComponentOrLiteral === 'boolean' ||
    ComponentOrLiteral == null
  ) {
    return ComponentOrLiteral;
  }
  return <ComponentOrLiteral {...props} />;
};

export const getRetinaImage = (images: string) =>
  images
    .split('|')
    .map((item, i) => `${item} ${i + 1}x`)
    .join(', ');

export function userOrDefault(
  user: UserResponse | string | {| error: string |},
): UserResponse {
  let actor: UserResponse;
  const notFound = {
    id: '!not-found',
    created_at: '',
    updated_at: '',
    data: { name: 'Unknown', profileImage: '' },
  };
  if (!user || typeof user === 'string' || typeof user.error === 'string') {
    actor = notFound;
  } else {
    //$FlowBug
    actor = (user: any);
  }
  return actor;
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// https://stackoverflow.com/a/6860916/2570866
export function generateRandomId() {
  // prettier-ignore
  return (S4()+S4()+"-"+S4()+"-"+S4()+"-"+S4()+"-"+S4()+S4()+S4());
}

function S4() {
  return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
}

export function dataTransferItemsHaveFiles(
  items: ?(DataTransferItem[]),
): boolean {
  if (!items || !items.length) {
    return false;
  }
  for (const item of items) {
    if (item.kind === 'file' || item.type === 'text/html') {
      return true;
    }
  }
  return false;
}

function getFileLikes(items) {
  const fileLikes = [];
  for (const item of items) {
    if (item.kind === 'file') {
      const file = item.getAsFile();
      if (file) {
        fileLikes.push(file);
      }
    }
  }
  return fileLikes;
}

export async function dataTransferItemsToFiles(
  items: ?(DataTransferItem[]),
): Promise<FileLike[]> {
  if (!items || !items.length) {
    return [];
  }

  const fileLikes = getFileLikes(items);
  // If there are files inside the DataTransferItem prefer those
  if (fileLikes.length) {
    return fileLikes;
  }

  // Otherwise extract images from html
  const blobPromises = [];
  const parser = new DOMParser();
  for (const item of items) {
    if (item.type === 'text/html') {
      blobPromises.push(
        new Promise((accept) => {
          item.getAsString(async (s) => {
            const doc = parser.parseFromString(s, 'text/html');
            const imageTags = doc.getElementsByTagName('img');

            const imagePromises = [];
            for (const tag of imageTags) {
              if (!tag.src) {
                continue;
              }
              imagePromises.push(
                (async () => {
                  let res;
                  try {
                    res = await fetch(tag.src);
                  } catch (e) {
                    return;
                  }
                  const contentType =
                    res.headers.get('Content-type') ||
                    'application/octet-stream';
                  const buf = await res.arrayBuffer();
                  const blob = new Blob([buf], { type: contentType });
                  fileLikes.push(blob);
                })(),
              );
            }
            await Promise.all(imagePromises);
            accept();
          });
        }),
      );
    }
  }
  await Promise.all(blobPromises);
  return fileLikes;
}

export function inputValueFromEvent(
  event?: ?SyntheticEvent<HTMLTextAreaElement> | ?Event,
): ?string {
  if (!event) {
    return;
  }
  let target;
  if (event.currentTarget) {
    target = event.currentTarget;
  } else {
    target = event.target;
  }
  // Trick flow into believing the target maybe has a value field
  const inputTarget: { value?: string } = (target: any);
  return inputTarget.value;
}

export function sanitizeURL(url: ?string): ?string {
  if (url == null) {
    return url;
  }

  const proto = URL(url).protocol;
  // allow http, https, ftp
  // IMPORTANT: Don't allow data: protocol because of:
  // <a href="data:text/html;base64,PHNjcmlwdD5hbGVydCgneHNzJyk7PC9zY3JpcHQ+" target="_blank">here</a>
  if (proto === 'https:' || proto === 'http:' || proto === 'ftp:') {
    return url;
  }
  return undefined;
}

const renderWord = (
  word: string,
  key: string,
  parentClass: string,
  onClickMention?: (word: string) => mixed,
  onClickHashtag?: (word: string) => mixed,
) => {
  if (onClickMention && word.includes('@')) {
    const mention = twitter.extractMentions(word);
    if (!mention.length) return word;

    return (
      <React.Fragment key={key}>
        {!word.startsWith(`@${mention[0]}`) &&
          word.slice(0, word.indexOf(mention[0]) - 1)}
        <a
          onClick={() => onClickMention && onClickMention(mention[0])}
          className={`${parentClass}__mention`}
        >
          @{mention[0]}
        </a>
        {!word.endsWith(mention[0]) &&
          word.slice(word.indexOf(mention[0]) + mention[0].length)}
      </React.Fragment>
    );
  } else if (onClickHashtag && word.includes('#')) {
    const hashtag = twitter.extractHashtags(word);
    if (!hashtag.length) return word;

    return (
      <React.Fragment key={key}>
        {!word.startsWith(`#${hashtag[0]}`) &&
          word.slice(0, word.indexOf(hashtag[0]) - 1)}
        <a
          onClick={() => onClickHashtag && onClickHashtag(hashtag[0])}
          className={`${parentClass}__hashtag`}
        >
          #{hashtag[0]}
        </a>
        {!word.endsWith(hashtag[0]) &&
          word.slice(word.indexOf(hashtag[0]) + hashtag[0].length)}
      </React.Fragment>
    );
  }
  if (anchorme.validate.url(word) || anchorme.validate.email(word)) {
    const link = anchorme(word, { list: true });
    if (
      link[0].protocol !== 'http://' &&
      link[0].protocol !== 'https://' &&
      link[0].protocol !== 'mailto:'
    ) {
      return word;
    }

    return (
      <a
        href={`${link[0].protocol}${link[0].encoded}`}
        className={`${parentClass}__link`}
        target="blank"
        rel="noopener"
        key={key}
      >
        {_truncate(link[0].encoded, { length: 33 })}
      </a>
    );
  }

  return word;
};

export const textRenderer = (
  text: string,
  parentClass: string,
  onClickMention?: (word: string) => mixed,
  onClickHashtag?: (word: string) => mixed,
) => {
  if (!text) return;

  return text
    .split(/\r\n|\r|\n/) // first break on line
    .map((line, i) =>
      line
        .split(' ') // break for each word
        .map((word, j) =>
          renderWord(
            word,
            `item-${i}-${j}`,
            parentClass,
            onClickMention,
            onClickHashtag,
          ),
        )
        .reduce((acc, elem) => (acc ? [acc, ' ', elem] : [elem])),
    )
    .reduce((acc, elem) => (acc ? [acc, '\n', elem] : [elem]));
};
