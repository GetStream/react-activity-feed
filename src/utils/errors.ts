import { StreamApiError, UR } from 'getstream';

export type NetworkRequestTypes =
  | 'get-user-info'
  | 'get-feed'
  | 'get-feed-next-page'
  | 'get-reactions-next-page'
  | 'get-notification-counts'
  | 'upload-image'
  | 'upload-file'
  | 'get-og'
  | 'add-activity'
  | 'delete-activity'
  | 'add-reaction'
  | 'delete-reaction'
  | 'add-child-reaction'
  | 'delete-child-reaction';

type ErrorDetail = UR & {
  activity?: UR;
  activity_id?: string;
  activityId?: string;
  feedGroup?: string;
  kind?: string;
  options?: UR;
  reaction?: UR;
  userId?: string;
};

export type ErrorHandler = (error: Error | StreamApiError, type: NetworkRequestTypes, details: ErrorDetail) => void;

export const handleError: ErrorHandler = (error, type, detail) => {
  console.warn(error, type, detail);
};

export const getErrorMessage: ErrorHandler = (error, type, detail) => {
  console.warn(error);

  if (!(error instanceof StreamApiError)) {
    return fallbackErrorMessage(error, type, detail);
  }

  const response = error.response;

  if (!response.status || !response.data || !response.data.detail) {
    return fallbackErrorMessage(error, type, detail);
  }
  const status = response.status;
  const text = response.data.detail;

  if (status >= 400 && status < 600) {
    return text;
  }

  return fallbackErrorMessage(error, type, detail);
};

export const fallbackErrorMessage: ErrorHandler = (_error, type, detail) => {
  let text = 'Something went wrong';
  let suffix = '';
  switch (type) {
    case 'get-user-info':
      text += ' when loading user info';
      break;
    case 'get-feed':
      text += ' when loading the feed';
      break;
    case 'get-feed-next-page':
      text += ' when loading the next page of the feed';
      break;
    case 'get-notification-counts':
      text += ' when loading your unread notification counts';
      break;
    case 'upload-image':
      text += ' when uploading your image';
      suffix = ' If it is, the image is probably too big';
      break;
    case 'add-activity':
      text += ' when submitting your post';
      break;
    case 'add-reaction':
    case 'add-child-reaction':
      text += ' when submitting your ' + detail.kind;
      break;
    case 'delete-reaction':
    case 'delete-child-reaction':
      text += ' when removing your ' + detail.kind;
      break;
    default:
  }

  return `${text}. Is your internet working?${suffix}`;
};
