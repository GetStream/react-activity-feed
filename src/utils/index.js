import React from 'react';
import URL from 'url-parse';
import _truncate from 'lodash/truncate';
import Dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import minMax from 'dayjs/plugin/minMax';
import * as linkify from 'linkifyjs';
import linkifyMention from 'linkifyjs/plugins/mention';

// 'linkifyjs/plugins/hashtag';
function linkifyHashtag(linkify) {
  const TT = linkify.scanner.TOKENS; // Text tokens
  const MultiToken = linkify.parser.TOKENS.Base; // Base Multi token class
  const S_START = linkify.parser.start;
  function HASHTAG(value) {
    this.v = value;
  }
  linkify.inherits(MultiToken, HASHTAG, { type: 'hashtag', isLink: true });
  const S_HASH = S_START.jump(TT.POUND);
  const S_HASHTAG = new linkify.parser.State(HASHTAG);

  S_HASH.on(TT.DOMAIN, S_HASHTAG);
  S_HASH.on(TT.UNDERSCORE, S_HASHTAG);
  S_HASH.on(TT.TLD, S_HASHTAG);

  // following lines are the diff from original implemention
  // add support for _ in hashtags
  S_HASH.on(TT.LOCALHOST, S_HASHTAG);
  S_HASHTAG.on(TT.UNDERSCORE, S_HASH);
}

linkifyMention(linkify);
linkifyHashtag(linkify);

Dayjs.extend(utc);
Dayjs.extend(minMax);

// import type { UserResponse } from 'getstream';

export function humanizeTimestamp(timestamp, tDateTimeParser) {
  let time;
  // Following calculation is based on assumption that tDateTimeParser()
  // either returns momentjs or dayjs object.

  // When timestamp doesn't have z at the end, we are supposed to take it as UTC time.
  // Ideally we need to adhere to RFC3339. Unfortunately this needs to be fixed on backend.
  if (typeof timestamp === 'string' && timestamp[timestamp.length - 1].toLowerCase() === 'z') {
    time = tDateTimeParser(timestamp);
  } else {
    time = tDateTimeParser(timestamp).add(Dayjs(timestamp).utcOffset(), 'minute'); // parse time as UTC
  }

  const now = tDateTimeParser();
  return time.from(now);
}

export const smartRender = (ElementOrComponentOrLiteral, props, fallback) => {
  if (ElementOrComponentOrLiteral === undefined) {
    ElementOrComponentOrLiteral = fallback;
  }
  if (React.isValidElement(ElementOrComponentOrLiteral)) {
    // Flow cast through any, to make flow believe it's a React.Element
    return ElementOrComponentOrLiteral;
  }

  // Flow cast through any to remove React.Element after previous check
  const ComponentOrLiteral = ElementOrComponentOrLiteral;
  if (
    typeof ComponentOrLiteral === 'string' ||
    typeof ComponentOrLiteral === 'number' ||
    typeof ComponentOrLiteral === 'boolean' ||
    ComponentOrLiteral == null
  ) {
    return ComponentOrLiteral;
  }
  return <ComponentOrLiteral {...props} />;
};

export function userOrDefault(user) {
  let actor;
  const notFound = {
    id: '!not-found',
    created_at: '',
    updated_at: '',
    data: { name: 'Unknown', profileImage: '' },
  };
  if (!user || typeof user === 'string' || typeof user.error === 'string') {
    actor = notFound;
  } else {
    actor = user;
  }
  return actor;
}

// https://stackoverflow.com/a/6860916/2570866
export function generateRandomId() {
  // prettier-ignore
  return (S4()+S4()+"-"+S4()+"-"+S4()+"-"+S4()+"-"+S4()+S4()+S4());
}

function S4() {
  return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
}

export function dataTransferItemsHaveFiles(items) {
  if (!items || !items.length) {
    return false;
  }
  for (const item of items) {
    if (item.kind === 'file' || item.type === 'text/html') {
      return true;
    }
  }
  return false;
}

function getFileLikes(items) {
  const fileLikes = [];
  for (const item of items) {
    if (item.kind === 'file') {
      const file = item.getAsFile();
      if (file) {
        fileLikes.push(file);
      }
    }
  }
  return fileLikes;
}

export async function dataTransferItemsToFiles(items) {
  if (!items || !items.length) {
    return [];
  }

  const fileLikes = getFileLikes(items);
  // If there are files inside the DataTransferItem prefer those
  if (fileLikes.length) {
    return fileLikes;
  }

  // Otherwise extract images from html
  const blobPromises = [];
  const parser = new DOMParser();
  for (const item of items) {
    if (item.type === 'text/html') {
      blobPromises.push(
        new Promise((accept) => {
          item.getAsString(async (s) => {
            const doc = parser.parseFromString(s, 'text/html');
            const imageTags = doc.getElementsByTagName('img');

            const imagePromises = [];
            for (const tag of imageTags) {
              if (!tag.src) {
                continue;
              }
              imagePromises.push(
                (async () => {
                  let res;
                  try {
                    res = await fetch(tag.src);
                  } catch (e) {
                    return;
                  }
                  const contentType = res.headers.get('Content-type') || 'application/octet-stream';
                  const buf = await res.arrayBuffer();
                  const blob = new Blob([buf], { type: contentType });
                  fileLikes.push(blob);
                })(),
              );
            }
            await Promise.all(imagePromises);
            accept();
          });
        }),
      );
    }
  }
  await Promise.all(blobPromises);
  return fileLikes;
}

export function inputValueFromEvent(event) {
  if (!event) {
    return;
  }
  let target;
  if (event.currentTarget) {
    target = event.currentTarget;
  } else {
    target = event.target;
  }
  // Trick flow into believing the target maybe has a value field
  const inputTarget = target;
  return inputTarget.value;
}

export function sanitizeURL(url) {
  if (url == null) {
    return url;
  }

  const proto = URL(url).protocol;
  // allow http, https, ftp
  // IMPORTANT: Don't allow data: protocol because of:
  // <a href="data:text/html;base64,PHNjcmlwdD5hbGVydCgneHNzJyk7PC9zY3JpcHQ+" target="_blank">here</a>
  if (proto === 'https:' || proto === 'http:' || proto === 'ftp:') {
    return url;
  }
  return undefined;
}

const renderWord = (word, key, parentClass, onClickMention, onClickHashtag) => {
  const [link] = linkify.find(word);
  if (!link) return word;

  const { type, value, href } = link;

  if (onClickMention && type === 'mention') {
    return (
      <React.Fragment key={key}>
        {!word.startsWith(value) && word.slice(0, word.indexOf(value))}
        <a onClick={() => onClickMention && onClickMention(value.substring(1))} className={`${parentClass}__mention`}>
          {value}
        </a>
        {!word.endsWith(value) && word.slice(word.indexOf(value) + value.length)}
      </React.Fragment>
    );
  }

  if (onClickHashtag && type === 'hashtag') {
    return (
      <React.Fragment key={key}>
        {!word.startsWith(value) && word.slice(0, word.indexOf(value))}
        <a onClick={() => onClickHashtag && onClickHashtag(value.substring(1))} className={`${parentClass}__hashtag`}>
          {value}
        </a>
        {!word.endsWith(value) && word.slice(word.indexOf(value) + value.length)}
      </React.Fragment>
    );
  }

  if (type === 'email' || type === 'url') {
    return (
      <a
        href={encodeURI(href)}
        className={`${parentClass}__link`}
        target="blank"
        data-testid="renderWord-hyperlink"
        rel="nofollow noreferrer noopener"
        key={key}
      >
        {type === 'email'
          ? value
          : _truncate(value.replace(/(http(s?):\/\/)?(www\.)?/, ''), {
              length: 33,
            })}
      </a>
    );
  }

  return word;
};

export const textRenderer = (text, parentClass, onClickMention, onClickHashtag) => {
  if (!text) return;

  return text
    .split(/\r\n|\r|\n/) // first break on line
    .map((line, i) =>
      line
        .split(' ') // break for each word
        .map((word, j) => renderWord(word, `item-${i}-${j}`, parentClass, onClickMention, onClickHashtag))
        .reduce((acc, elem) => (acc ? [acc, ' ', elem] : [elem])),
    )
    .reduce((acc, elem) => (acc ? [acc, '\n', elem] : [elem]));
};

/**
 *
 * @param {string | undefined} url
 */
export const trimURL = (url) =>
  url
    ?.replace(/^(?:https?:\/\/)?(?:www\.)?/i, '')
    .split('/')
    .shift();
