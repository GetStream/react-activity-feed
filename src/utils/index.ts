import React from 'react';
import URL from 'url-parse';
import Dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import minMax from 'dayjs/plugin/minMax';
import { EnrichedUser, UR } from 'getstream';
import { TDateTimeParser } from '../i18n/Streami18n';

Dayjs.extend(utc);
Dayjs.extend(minMax);

// import type { UserResponse } from 'getstream';

export function humanizeTimestamp(timestamp: string | Dayjs.Dayjs, tDateTimeParser: TDateTimeParser) {
  let time;
  // Following calculation is based on assumption that tDateTimeParser()
  // either returns momentjs or dayjs object.

  // When timestamp doesn't have z at the end, we are supposed to take it as UTC time.
  // Ideally we need to adhere to RFC3339. Unfortunately this needs to be fixed on backend.
  if (typeof timestamp === 'string' && timestamp[timestamp.length - 1].toLowerCase() === 'z') {
    time = tDateTimeParser(timestamp);
  } else {
    time = tDateTimeParser(timestamp).add(Dayjs(timestamp).utcOffset(), 'minute'); // parse time as UTC
  }

  const now = tDateTimeParser();
  return time.from(now);
}

type ErrorUser = { error: string };
function isErrorUser(user: unknown | ErrorUser): user is ErrorUser {
  return !!user && typeof (user as ErrorUser).error === 'string';
}

export function userOrDefault<T extends UR = UR>(user?: EnrichedUser<T> | string | { error: string } | null) {
  if (!user || typeof user === 'string' || isErrorUser(user))
    return {
      id: '!not-found',
      created_at: '',
      updated_at: '',
      data: { name: 'Unknown', profileImage: '' },
    };

  return user;
}

// https://stackoverflow.com/a/6860916/2570866
export function generateRandomId() {
  // prettier-ignore
  return (S4()+S4()+"-"+S4()+"-"+S4()+"-"+S4()+"-"+S4()+S4()+S4());
}

function S4() {
  return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
}

export function dataTransferItemsHaveFiles(items?: DataTransferItemList) {
  if (!items || !items.length) return false;

  for (let i = 0; i < items.length; i += 1) {
    const item = items.item(i);
    if (item.kind === 'file' || item.type === 'text/html') return true;
  }

  return false;
}

function getFileLikes(items: DataTransferItemList) {
  const fileLikes: Array<Blob | File> = [];
  for (let i = 0; i < items.length; i += 1) {
    const item = items.item(i);
    if (item.kind === 'file') {
      const file = item.getAsFile();
      if (file) fileLikes.push(file);
    }
  }
  return fileLikes;
}

export async function dataTransferItemsToFiles(items?: DataTransferItemList) {
  if (!items || !items.length) return [];

  const fileLikes = getFileLikes(items);
  // If there are files inside the DataTransferItem prefer those
  if (fileLikes.length) return fileLikes;

  // Otherwise extract images from html
  const blobPromises = [];
  const parser = new DOMParser();
  for (let i = 0; i < items.length; i += 1) {
    const item = items.item(i);
    if (item.type === 'text/html') {
      blobPromises.push(
        new Promise((accept) => {
          item.getAsString(async (s) => {
            const doc = parser.parseFromString(s, 'text/html');
            const imageTags = doc.getElementsByTagName('img');

            const imagePromises = [];
            for (let j = 0; j < imageTags.length; j++) {
              const tag = imageTags[j];
              if (!tag.src) {
                continue;
              }
              imagePromises.push(
                (async () => {
                  let res;
                  try {
                    res = await fetch(tag.src);
                  } catch (e) {
                    return;
                  }
                  const contentType = res.headers.get('Content-type') || 'application/octet-stream';
                  const buf = await res.arrayBuffer();
                  const blob = new Blob([buf], { type: contentType });
                  fileLikes.push(blob);
                })(),
              );
            }
            await Promise.all(imagePromises);
            accept(true);
          });
        }),
      );
    }
  }
  await Promise.all(blobPromises);
  return fileLikes;
}

export function inputValueFromEvent(event?: React.SyntheticEvent<HTMLInputElement>) {
  const target = (event?.currentTarget || event?.target) as HTMLInputElement;
  return target?.value;
}

export function sanitizeURL(url?: string) {
  if (!url) return url;

  const proto = URL(url).protocol;
  // allow http, https, ftp
  // IMPORTANT: Don't allow data: protocol because of:
  // <a href="data:text/html;base64,PHNjcmlwdD5hbGVydCgneHNzJyk7PC9zY3JpcHQ+" target="_blank">here</a>
  if (proto === 'https:' || proto === 'http:' || proto === 'ftp:') {
    return url;
  }
  return undefined;
}

/**
 *
 * @param {string | undefined} url
 */
export const trimURL = (url?: string) =>
  url
    ?.replace(/^(?:https?:\/\/)?(?:www\.)?/i, '')
    .split('/')
    .shift();

export * from './textRenderer';
export * from './smartRender';
