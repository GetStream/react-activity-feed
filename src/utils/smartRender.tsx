import React, { createElement, ReactElement, ElementType, ComponentType } from 'react';
import { UR } from 'getstream';

type ElementOrComponentOrLiteralType<T> =
  | string
  | number
  | boolean
  | null
  | ReactElement<T>
  | ElementType<T>
  | ComponentType<T>;

export function smartRender<T extends UR = UR>(
  ElementOrComponentOrLiteral?: ElementOrComponentOrLiteralType<T>,
  props?: T,
  fallback?: ElementOrComponentOrLiteralType<T>,
) {
  const RenderComponent = ElementOrComponentOrLiteral === undefined ? fallback : ElementOrComponentOrLiteral;

  if (React.isValidElement(RenderComponent)) return RenderComponent;

  if (
    typeof RenderComponent === 'string' ||
    typeof RenderComponent === 'number' ||
    typeof RenderComponent === 'boolean' ||
    RenderComponent == null
  ) {
    return RenderComponent;
  }

  return createElement<T>(RenderComponent, props);
}
