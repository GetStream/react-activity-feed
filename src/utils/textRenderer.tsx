import React from 'react';
import _truncate from 'lodash/truncate';
import * as linkify from 'linkifyjs';
// @ts-expect-error
import linkifyMention from 'linkifyjs/plugins/mention';

// 'linkifyjs/plugins/hashtag';
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function linkifyHashtag(linkify: any) {
  const TT = linkify.scanner.TOKENS; // Text tokens
  const MultiToken = linkify.parser.TOKENS.Base; // Base Multi token class
  const S_START = linkify.parser.start;
  function HASHTAG(value: unknown) {
    // @ts-expect-error
    this.v = value;
  }
  linkify.inherits(MultiToken, HASHTAG, { type: 'hashtag', isLink: true });
  const S_HASH = S_START.jump(TT.POUND);
  const S_HASHTAG = new linkify.parser.State(HASHTAG);

  S_HASH.on(TT.DOMAIN, S_HASHTAG);
  S_HASH.on(TT.UNDERSCORE, S_HASHTAG);
  S_HASH.on(TT.TLD, S_HASHTAG);

  // following lines are the diff from original implemention
  // add support for _ in hashtags
  S_HASH.on(TT.LOCALHOST, S_HASHTAG);
  S_HASHTAG.on(TT.UNDERSCORE, S_HASH);
}

linkifyMention(linkify);
linkifyHashtag(linkify);

type ClickCallback = (word: string) => void;
type Word = string | JSX.Element;
type WordArray = Array<Word | Word[] | WordArray>;
type WordArrayArray = Array<WordArray | Word | WordArrayArray>;

type CustomAnchorProps = {
  type: 'mention' | 'hashtag';
  value: string;
  word: string;
  clickCallback?: ClickCallback;
  parentClass?: string;
};

const CustomAnchor = ({ type, word, parentClass, value, clickCallback = () => {} }: CustomAnchorProps) => (
  <React.Fragment>
    {!word.startsWith(value) && word.slice(0, word.indexOf(value))}
    <a onClick={() => clickCallback(value.substring(1))} className={`${parentClass}__${type}`}>
      {value}
    </a>
    {!word.endsWith(value) && word.slice(word.indexOf(value) + value.length)}
  </React.Fragment>
);

const renderWord = (
  word: string,
  key: string,
  parentClass?: string,
  onClickMention?: ClickCallback,
  onClickHashtag?: ClickCallback,
): Word => {
  const [link] = linkify.find(word);
  if (!link) return word;

  const { type, value, href } = link;

  if (onClickMention && type === 'mention') {
    return (
      <CustomAnchor
        key={key}
        type={type}
        value={value}
        word={word}
        clickCallback={onClickMention}
        parentClass={parentClass}
      />
    );
  }

  if (onClickHashtag && type === 'hashtag') {
    return (
      <CustomAnchor
        key={key}
        type={type}
        value={value}
        word={word}
        clickCallback={onClickHashtag}
        parentClass={parentClass}
      />
    );
  }

  if (type === 'email' || type === 'url') {
    return (
      <a
        href={encodeURI(href)}
        className={`${parentClass}__link`}
        target="blank"
        data-testid="renderWord-hyperlink"
        rel="nofollow noreferrer noopener"
        key={key}
      >
        {type === 'email' ? value : _truncate(value.replace(/(http(s?):\/\/)?(www\.)?/, ''), { length: 33 })}
      </a>
    );
  }

  return word;
};

export const textRenderer = (
  text?: string,
  parentClass?: string,
  onClickMention?: ClickCallback,
  onClickHashtag?: ClickCallback,
) => {
  if (!text) return <></>;

  const renderedText = text
    .split(/\r\n|\r|\n/) // first break on line
    .map((line, i) =>
      line
        .split(' ') // break for each word
        .map((word, j) => renderWord(word, `item-${i}-${j}`, parentClass, onClickMention, onClickHashtag))
        .reduce<WordArray>((acc, elem) => (acc.length ? [acc, ' ', elem] : [elem]), []),
    )
    .reduce<WordArrayArray>((acc, elem) => (acc.length ? [acc, '\n', elem] : [elem]), []);

  return <>{renderedText}</>;
};
