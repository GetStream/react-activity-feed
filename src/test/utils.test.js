// @flow
import { textRenderer } from '../utils';
import renderer from 'react-test-renderer';

describe('utils -> textRenderer', () => {
  const onMentionClick = (word) => word;

  const textRendererWithCallbacks = (text) =>
    textRenderer(text, '', onMentionClick, (word) => word);

  it('is renders @ without callback func correctly ', () => {
    const tree = renderer
      .create(
        textRenderer(
          "@allison's @allison! @allison? (@allison) [@allison] @allison; @allison: @allison, @allison @allison_allison",
          '',
        ),
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('is renders @ with callback func correctly ', () => {
    const tree = renderer
      .create(
        textRendererWithCallbacks(
          "@allison's @allison! @allison? (@allison) [@allison] @allison; @allison: @allison, @allison @allison_allison",
        ),
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('is renders # without callback func correctly', () => {
    const tree = renderer
      .create(
        textRenderer(
          "#stream's #stream! #stream? (#stream) [#stream] #stream; #stream: #stream, #stream #stream_stream",
          '',
        ),
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('is renders # with callback func correctly', () => {
    const tree = renderer
      .create(
        textRendererWithCallbacks(
          "#stream's #stream! #stream? (#stream) [#stream] #stream; #stream: #stream, #stream #stream_stream",
        ),
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
