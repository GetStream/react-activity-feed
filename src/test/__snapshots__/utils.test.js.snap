// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils -> textRenderer is renders # with callback func correctly 1`] = `
Array [
  <a
    className="__hashtag"
    onClick={[Function]}
  >
    #
    stream
  </a>,
  "'s",
  " ",
  <a
    className="__hashtag"
    onClick={[Function]}
  >
    #
    stream
  </a>,
  "!",
  " ",
  <a
    className="__hashtag"
    onClick={[Function]}
  >
    #
    stream
  </a>,
  "?",
  " ",
  "(",
  <a
    className="__hashtag"
    onClick={[Function]}
  >
    #
    stream
  </a>,
  ")",
  " ",
  "[",
  <a
    className="__hashtag"
    onClick={[Function]}
  >
    #
    stream
  </a>,
  "]",
  " ",
  <a
    className="__hashtag"
    onClick={[Function]}
  >
    #
    stream
  </a>,
  ";",
  " ",
  <a
    className="__hashtag"
    onClick={[Function]}
  >
    #
    stream
  </a>,
  ":",
  " ",
  <a
    className="__hashtag"
    onClick={[Function]}
  >
    #
    stream
  </a>,
  ",",
  " ",
  <a
    className="__hashtag"
    onClick={[Function]}
  >
    #
    stream
  </a>,
  " ",
  <a
    className="__hashtag"
    onClick={[Function]}
  >
    #
    stream_stream
  </a>,
]
`;

exports[`utils -> textRenderer is renders # without callback func correctly 1`] = `
Array [
  "#stream's",
  " ",
  "#stream!",
  " ",
  "#stream?",
  " ",
  "(#stream)",
  " ",
  "[#stream]",
  " ",
  "#stream;",
  " ",
  "#stream:",
  " ",
  "#stream,",
  " ",
  "#stream",
  " ",
  "#stream_stream",
]
`;

exports[`utils -> textRenderer is renders @ with callback func correctly  1`] = `
Array [
  <a
    className="__mention"
    onClick={[Function]}
  >
    @
    allison
  </a>,
  "'s",
  " ",
  <a
    className="__mention"
    onClick={[Function]}
  >
    @
    allison
  </a>,
  "!",
  " ",
  <a
    className="__mention"
    onClick={[Function]}
  >
    @
    allison
  </a>,
  "?",
  " ",
  "(",
  <a
    className="__mention"
    onClick={[Function]}
  >
    @
    allison
  </a>,
  ")",
  " ",
  "[",
  <a
    className="__mention"
    onClick={[Function]}
  >
    @
    allison
  </a>,
  "]",
  " ",
  <a
    className="__mention"
    onClick={[Function]}
  >
    @
    allison
  </a>,
  ";",
  " ",
  <a
    className="__mention"
    onClick={[Function]}
  >
    @
    allison
  </a>,
  ":",
  " ",
  <a
    className="__mention"
    onClick={[Function]}
  >
    @
    allison
  </a>,
  ",",
  " ",
  <a
    className="__mention"
    onClick={[Function]}
  >
    @
    allison
  </a>,
  " ",
  <a
    className="__mention"
    onClick={[Function]}
  >
    @
    allison_allison
  </a>,
]
`;

exports[`utils -> textRenderer is renders @ without callback func correctly  1`] = `
Array [
  "@allison's",
  " ",
  "@allison!",
  " ",
  "@allison?",
  " ",
  "(@allison)",
  " ",
  "[@allison]",
  " ",
  "@allison;",
  " ",
  "@allison:",
  " ",
  "@allison,",
  " ",
  "@allison",
  " ",
  "@allison_allison",
]
`;
